### ISSUE TITLE
Robolectric crashes with new VectorDrawableCompat

### ISSUE THREAD

**Issue by hzsweers** [robolectric/robolectric#2315]
When AppCompat tries to create a drawable from a given vector, Robolectric tests fail with the following exception:

```
Caused by: org.xmlpull.v1.XmlPullParserException: XML file build/intermediates/res/merged/world/debug/drawable-anydpi-v21/abc_ic_ab_back_material.xml line #-1 (sorry, not yet implemented): invalid drawable tag vector
	at android.graphics.drawable.Drawable.createFromXmlInner(Drawable.java:897)
	at android.graphics.drawable.Drawable.createFromXml(Drawable.java:837)
	at android.content.res.Resources.loadDrawable(Resources.java:2087)
	... 57 more
```

Happens regardless of whether or not support vectors are enabled as specified here: http://android-developers.blogspot.com/2016/02/android-support-library-232.html

```groovy
// Gradle Plugin 2.0+  
 android {  
   defaultConfig {  
     vectorDrawables.useSupportLibrary = true  
    }  
 }  
```

Relevant exception higher up:
```
Caused by: android.content.res.Resources$NotFoundException: File build/intermediates/res/merged/world/debug/drawable/abc_ic_ab_back_material.xml from drawable resource ID #0x7f020016
	at android.content.res.Resources.loadDrawable(Resources.java:2091)
	at org.robolectric.util.ReflectionHelpers.callInstanceMethod(ReflectionHelpers.java:195)
	at org.robolectric.internal.Shadow.directlyOn(Shadow.java:57)
	at org.robolectric.shadows.ShadowResources.loadDrawable(ShadowResources.java:518)
	at android.content.res.Resources.loadDrawable(Resources.java)
	at android.content.res.Resources.getDrawable(Resources.java:695)
	at android.support.v4.content.ContextCompat.getDrawable(ContextCompat.java:323)
	at android.support.v7.widget.AppCompatDrawableManager.getDrawable(AppCompatDrawableManager.java:180)
	at android.support.v7.widget.AppCompatDrawableManager.getDrawable(AppCompatDrawableManager.java:173)
	at android.support.v7.widget.TintTypedArray.getDrawable(TintTypedArray.java:60)
	at android.support.v7.widget.Toolbar.__constructor__(Toolbar.java:254)
	at android.support.v7.widget.Toolbar.<init>(Toolbar.java)
	at android.view.LayoutInflater.createView(LayoutInflater.java:594)
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:696)
	at android.view.LayoutInflater.rInflate(LayoutInflater.java:755)
	at android.view.LayoutInflater.rInflate(LayoutInflater.java:758)
	at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
	at uk.co.chrisjenx.calligraphy.CalligraphyLayoutInflater.inflate(CalligraphyLayoutInflater.java:60)
	at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
	at android.view.LayoutInflater.inflate(LayoutInflater.java:353)
	at android.support.v7.app.AppCompatDelegateImplV7.createSubDecor(AppCompatDelegateImplV7.java:379)
	at android.support.v7.app.AppCompatDelegateImplV7.ensureSubDecor(AppCompatDelegateImplV7.java:300)
	at android.support.v7.app.AppCompatDelegateImplV7.initWindowDecorActionBar(AppCompatDelegateImplV7.java:169)
	at android.support.v7.app.AppCompatDelegateImplBase.getSupportActionBar(AppCompatDelegateImplBase.java:88)
	at android.support.v7.app.AppCompatActivity.getSupportActionBar(AppCompatActivity.java:99)
```

AppCompat does some working around for vector drawables via its `ResourcesWrapper` and `ContextWrapper` classes, I think the solution for this is that Robolectric needs to work with this. See the bottom of https://medium.com/@chrisbanes/appcompat-v23-2-age-of-the-vectors-91cbafa87c88#.hq4q2olt3

**Comment by jongerrish** [comment]
I _think_ this is fixed at head.

Can you try with a snapshot build?

**Comment by hzsweers** [comment]
I tried with both 3.0 and the 3.2-SNAPSHOT, neither worked :(

**Comment by jongerrish** [comment]
You mean 3.1-SNAPSHOT?

Can you post a simple test that exposes this issue?

**Comment by hzsweers** [comment]
oops, yeah that was a typo. Still happens with the 3.**1**-SNAPSHOT. I'll try to put together a branch with an example

**Comment by hzsweers** [comment]
@jongerrish here you go: https://github.com/hzsweers/RobolectricVectorDrawableCompatBug

**Comment by jongerrish** [comment]
You have @Config(sdk = 16) Vectors are not available till 21 no?

**Comment by hzsweers** [comment]
Read the links I included at the top. AppCompat 23.2.0 introduced `VectorDrawableCompat` and `AnimatedVectorDrawableCompat`, backporting them to API 7 and 11 respectively.

**Comment by hzsweers** [comment]
@jongerrish ping. I think this is going to be a blocker for anyone wanting to use the new support library updates going forward

**Comment by bjenning04** [comment]
My team is also running into this issue with VectorDrawableCompat. Here is our stack trace:
```
android.view.InflateException: XML file build/intermediates/res/merged/debug/layout/abc_screen_toolbar.xml line #-1 (sorry, not yet implemented): Error inflating class <unknown>
        at android.view.LayoutInflater.createView(LayoutInflater.java:620)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:696)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:755)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:758)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:353)
        at android.support.v7.app.AppCompatDelegateImplV7.createSubDecor(AppCompatDelegateImplV7.java:379)
        at android.support.v7.app.AppCompatDelegateImplV7.ensureSubDecor(AppCompatDelegateImplV7.java:300)
        at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java:264)
        at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java:129)
        at com.cerner.careaware.dialer.activity.CallActivity.onCreate(CallActivity.java:32)
        at android.app.Activity.performCreate(Activity.java:5133)
        at org.robolectric.util.ReflectionHelpers.callInstanceMethod(ReflectionHelpers.java:195)
        at org.robolectric.util.ActivityController$1.run(ActivityController.java:127)
        at org.robolectric.shadows.ShadowLooper.runPaused(ShadowLooper.java:340)
        at org.robolectric.shadows.CoreShadowsAdapter$2.runPaused(CoreShadowsAdapter.java:45)
        at org.robolectric.util.ActivityController.create(ActivityController.java:123)
        at org.robolectric.util.ActivityController.create(ActivityController.java:134)
        at com.cerner.careaware.dialer.view.CallBannerTest.setup(CallBannerTest.java:37)

        Caused by:
        java.lang.reflect.InvocationTargetException
            at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
            at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
            at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
            at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
            at android.view.LayoutInflater.$$robo$$createView(LayoutInflater.java:594)
            at android.view.LayoutInflater.createView(LayoutInflater.java)
            at android.view.LayoutInflater.$$robo$$createViewFromTag(LayoutInflater.java:696)
            at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java)
            at android.view.LayoutInflater.$$robo$$rInflate(LayoutInflater.java:755)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java)
            at android.view.LayoutInflater.$$robo$$rInflate(LayoutInflater.java:758)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java)
            at android.view.LayoutInflater.$$robo$$inflate(LayoutInflater.java:492)
            at android.view.LayoutInflater.inflate(LayoutInflater.java)
            at android.view.LayoutInflater.$$robo$$inflate(LayoutInflater.java:397)
            at android.view.LayoutInflater.inflate(LayoutInflater.java)
            at android.view.LayoutInflater.$$robo$$inflate(LayoutInflater.java:353)
            at android.view.LayoutInflater.inflate(LayoutInflater.java)
            at android.support.v7.app.AppCompatDelegateImplV7.$$robo$$createSubDecor(AppCompatDelegateImplV7.java:379)
            at android.support.v7.app.AppCompatDelegateImplV7.createSubDecor(AppCompatDelegateImplV7.java)
            at android.support.v7.app.AppCompatDelegateImplV7.$$robo$$ensureSubDecor(AppCompatDelegateImplV7.java:300)
            at android.support.v7.app.AppCompatDelegateImplV7.ensureSubDecor(AppCompatDelegateImplV7.java)
            at android.support.v7.app.AppCompatDelegateImplV7.$$robo$$setContentView(AppCompatDelegateImplV7.java:264)
            at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java)
            at android.support.v7.app.AppCompatActivity.$$robo$$setContentView(AppCompatActivity.java:129)
            at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java)
            at com.cerner.careaware.dialer.activity.CallActivity.onCreate(CallActivity.java:32)
            at android.app.Activity.$$robo$$performCreate(Activity.java:5133)
            at android.app.Activity.performCreate(Activity.java)
            at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
            at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
            at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
            at java.lang.reflect.Method.invoke(Method.java:606)
            ... 7 more

            Caused by:
            android.content.res.Resources$NotFoundException: File build/intermediates/res/merged/debug/drawable-anydpi-v21/abc_ic_ab_back_material.xml from drawable resource ID #0x7f020016
                at android.content.res.Resources.loadDrawable(Resources.java:2091)
                at org.robolectric.util.ReflectionHelpers.callInstanceMethod(ReflectionHelpers.java:195)
                at org.robolectric.internal.Shadow.directlyOn(Shadow.java:57)
                at org.robolectric.shadows.ShadowResources.loadDrawable(ShadowResources.java:518)
                at android.content.res.Resources.loadDrawable(Resources.java)
                at android.content.res.Resources.getDrawable(Resources.java:695)
                at android.support.v4.content.ContextCompat.getDrawable(ContextCompat.java:323)
                at android.support.v7.widget.AppCompatDrawableManager.getDrawable(AppCompatDrawableManager.java:180)
                at android.support.v7.widget.AppCompatDrawableManager.getDrawable(AppCompatDrawableManager.java:173)
                at android.support.v7.widget.TintTypedArray.getDrawable(TintTypedArray.java:60)
                at android.support.v7.widget.Toolbar.__constructor__(Toolbar.java:254)
                at android.support.v7.widget.Toolbar.<init>(Toolbar.java)
                at android.view.LayoutInflater.createView(LayoutInflater.java:594)
                at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:696)
                at android.view.LayoutInflater.rInflate(LayoutInflater.java:755)
                at android.view.LayoutInflater.rInflate(LayoutInflater.java:758)
                at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
                at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
                at android.view.LayoutInflater.inflate(LayoutInflater.java:353)
                at android.support.v7.app.AppCompatDelegateImplV7.createSubDecor(AppCompatDelegateImplV7.java:379)
                at android.support.v7.app.AppCompatDelegateImplV7.ensureSubDecor(AppCompatDelegateImplV7.java:300)
                at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java:264)
                at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java:129)
                at com.cerner.careaware.dialer.activity.CallActivity.onCreate(CallActivity.java:32)
                at android.app.Activity.performCreate(Activity.java:5133)
                ... 7 more

                Caused by:
                org.xmlpull.v1.XmlPullParserException: XML file build/intermediates/res/merged/debug/drawable-anydpi-v21/abc_ic_ab_back_material.xml line #-1 (sorry, not yet implemented): invalid drawable tag vector
                    at android.graphics.drawable.Drawable.createFromXmlInner(Drawable.java:897)
                    at android.graphics.drawable.Drawable.createFromXml(Drawable.java:837)
                    at android.content.res.Resources.loadDrawable(Resources.java:2087)
                    ... 31 more
```

**Comment by jongerrish** [comment]
So the code is calling down into the framework code for APIs < 21:-

https://github.com/android/platform_frameworks_base/blob/kitkat-mr2.2-release/graphics/java/android/graphics/drawable/Drawable.java

Which doesn't support <vector> hence that it blows up.

I suspect that this is due to Robolectric's shadow implementations (ShadowContextWrapper, ShadowResources) overriding the tricks performed by appcompat (see here:- https://medium.com/@chrisbanes/appcompat-v23-2-age-of-the-vectors-91cbafa87c88#.1vcokofeb)

**Comment by hzsweers** [comment]
Agreed, fairly certain that's the case. Do you think that can be worked around?

**Comment by jongerrish** [comment]
Yeah, I'm pretty sure we can solve it by removing shadow code.

ShadowContextWrapper shouldn't even exist as the framework code just
delegates through.

In fact, it's just wrong as it redirects calls to the application when it
should be the delegate.

**Comment by hzsweers** [comment]
Do you think you'll be able to look into it? I gave it a shot but I'm not sure it's actually properly routing through `ShadowContextWrapper` judging by the stacktrace. It seems like the problem lies deeper, like it should detect within Resources.getDrawable if it should delegate to AppCompat's ResourceWrapper.

```
at android.content.res.Resources.loadDrawable(Resources.java:1918)
	at org.robolectric.util.ReflectionHelpers.callInstanceMethod(ReflectionHelpers.java:195)
	at org.robolectric.internal.Shadow.directlyOn(Shadow.java:44)
	at org.robolectric.shadows.ShadowResources.loadDrawable(ShadowResources.java:505)
	at android.content.res.Resources.loadDrawable(Resources.java)
	at android.content.res.Resources.getDrawable(Resources.java:659)
	at android.support.v7.widget.TintResources.superGetDrawable(TintResources.java:48)
```

**Comment by jongerrish** [comment]
I'm looking at it now.

**Comment by hzsweers** [comment]
Awesome!

**Comment by jaredsburrows** [comment]
@hzsweers Have you tried the latest app compat `23.2.1`? Are you still using the `2.+` plugin? Have you tried with `1.5.0`?

**Comment by jongerrish** [comment]
Can you check again with the latest snapshot? I've submitted a bunch of PRs
recently that removed a lot of shadow code from ContextWrapper which was
interfering with the hacks AppCompat uses to support vectors on pre 21.

**Comment by jaredsburrows** [comment]
@jongerrish I actually have never tried it before. I do not have test project setup right now. I just wanted to make sure the support if available so the unit tests do not suffer. I want to make a proposal that we move from PNGs to SVGs. 

Side question: Is there a way Robolectric can start being released more often? I would really like to use API 22 and 23 support for SDKs and Android applications.

**Comment by hzsweers** [comment]
Yup the latest master works for vectordrawablecompat now, thanks!

Agreed with @jaredsburrows about releases. Need I to rely on snapshots won't work for us, and it's been some time since the last release. Also, powermock has been broken on master for months and the issue for it hasn't been acknowledged. Is there any plan to resolve that? It's somewhat essential when dealing with settings, BuildConfig, support library Compat utilities, etc.

**Comment by jaredsburrows** [comment]
@hzsweers Thanks for replying. We can probably now close this issue right? Also, my main concern is switching to snapshots that make break the CI builds. I think more incremental releases will help with that as well as prevent duplicated issues from being created.

**Comment by hzsweers** [comment]
Sure, I can close the issue. I'd like to get a response from @jongerrish while he's here though first if possible since the powermock issue seems to be buried :/

**Comment by jaredsburrows** [comment]
@hzsweers Yeah definitely. Based on design, I try to not mock and if I do, I try to not mock any static or final classes and simply use `mockito` instead.

**Comment by jongerrish** [comment]
@jaredsburrows - My focus right now is improving head to support internal customers but I believe @mag has taken on pushing out the 3.1 release. I'm definitely in favor of a regular release cycle, one for at least every new Android sdk version or major Robolectic feature and regular point releases with bug fixes. I'd be supportive if anyone wanted to take that on.

@hzsweers - Thanks for verifying the AppCompat fix, that makes me very happy :-) In general I want to start deleting unnecessary shadow code, a lot of that is a vestige of Robolectric 1.0 which didn't use real AOSP code, I want to lean on real Android where possible to improve fidelity. Personally I agree with @jaredsburrows in that Mockito and design for testability you can avoid Power Mock but for projects that don't do this, I think Power Mock is better than writing custom shadows for your own code. With this in mind, I'm actually in favor of restricting Shadowing to only Android Framework classes and telling users to prefer Power Mock for testing static code. I'd prefer Robolectric just focus on simulating Android and be good at that and using PowerMock / Mockito were appropriate. What do you think?

**Comment by hzsweers** [comment]
I would agree with that logic, and definitely in favor relying on the framework directly where possible. Ideally a testing framework should be as transparent of a layer as possible, and avoid situations where the developer is really testing the test framework rather than the real thing.

For Power Mock, I think the issue is much deeper and I think I maybe should've been clearer in my earlier comment. The issue I was referencing is #2208, which appears to break any test that attempts to use Power Mock, even for personal tests and not necessarily Robolectric ones or in place of shadows. I'd love to prefer referring people to Power Mock for testing static code, but unfortunately it seems that current master prevents this.

**Comment by jaredsburrows** [comment]
@jongerrish @hzsweers I agree with shadows for Android only and mocking classes for app implementations. I try to avoid static classes/final classes that cant be tested and use `mockito` as a last resort. I do believe that Robolectric supports PowerMockito out of the box. 

See PowerMockito here: https://github.com/robolectric/robolectric/wiki/Using-PowerMock and the issue originally from here: https://github.com/robolectric/robolectric/issues/1058.

**Comment by hzsweers** [comment]
> I do believe that Robolectric supports PowerMockito out of the box.

My point is that this no longer seems to be the case :/

**Comment by jaredsburrows** [comment]
@hzsweers Are you using PowerMock or PowerMockito? PowerMock has never worked for me but PowerMockito has. 

Please see this example here: https://github.com/WonderCsabo/deckard-gradle.

**Comment by hzsweers** [comment]
@jaredsburrows I don't see either used in that project, could you maybe link me a specific file?

**Comment by jaredsburrows** [comment]
@hzsweers Sorry about that.
It is here: https://github.com/WonderCsabo/deckard-gradle/tree/PowerMock

Dependencies: https://github.com/WonderCsabo/deckard-gradle/blob/PowerMock/build.gradle#L52

Check the test here: https://github.com/WonderCsabo/deckard-gradle/blob/PowerMock/src/test/java/com/example/activity/DeckardActivityTest.java#L21

**Comment by hzsweers** [comment]
Also we're using PowerMockito

**Comment by jaredsburrows** [comment]
@hzsweers So is @wondercsabo in his example.

**Comment by hzsweers** [comment]
Right, but he's not using the master snapshot, which is where power mock is actually broken.

**Comment by jaredsburrows** [comment]
So your issue is vector graphics + powermockito and not just powermockito?

**Comment by hzsweers** [comment]
It *was* both, now vector drawable compat is fixed on master but we can't use the snapshots to take advantage of that because powermockito is broken on master as well. One kind of led to another and I opted to point that out here since the other issue had been missed.

Before on master snapshots: vector drawable compat **and** powermockito both were broken.

Now on master snapshots: powermockito is still broken.

**Comment by jaredsburrows** [comment]
Alright. Thanks for the clarification.

Jared Burrows

*jaredsburrows@gmail.com <jaredsburrows@gmail.com>*
My Resume: https://www.linkedin.com/in/jaredsburrows

**Comment by hzsweers** [comment]
VectorDrawableCompat is working now

**Comment by xwei-handy** [comment]
Hi, which version will this fix be released to?

**Comment by jongerrish** [comment]
The next release (3.1.2), it's top priority.

**Comment by hzsweers** [comment]
Wait, 3.1 didn't fix this?

**Comment by jaredsburrows** [comment]
@jongerrish +1 on the quick releases.

**Comment by jongerrish** [comment]
I don't know. It depends on the version of AppCompat, the one I was testing
triggers that catastrophic failure. If yours doesn't then that's great. But
the bug still remains and I need to fix it.

**Comment by MiguelLavigne** [comment]
Hi, I see that this issue has been closed but I wanted to point out something in case someone can help.

Is the solution that's been discussed here suppose to fix this running when running against sdk version 16 using VectorDrawableCompat?

I tried the project that @hzsweers linked  up in the thread https://github.com/hzsweers/RobolectricVectorDrawableCompatBug and I updated it to reference the latest robolectric 3.2-SNAPSHOT and this issue is still happening if run against SDK 16.

Can anyone clarify what should be expect?
Thank you!

**Comment by jongerrish** [comment]
Yeah, that's the bug indeed! 

This is a duplicate I guess there is another issue to track the broken vector drawable for compat on API < 21

**Comment by jongerrish** [comment]
We'll target it to 3.1.2

**Comment by MiguelLavigne** [comment]
Hey @jongerrish from your answer I take it the issue isn't fixed yet.  I think my confusion came from the previous discussion.  I thought it had already been fixed that's why I gave the 3.2-SNAPSHOT a run.

Thanks for clarifying this for me.

**Comment by MiguelLavigne** [comment]
You mentioned there was another issue tracking the progress for the `vector drawable for compat on API < 21`.  Would you be able to link me this issue?

**Comment by jongerrish** [comment]
Duplicate of: https://github.com/robolectric/robolectric/issues/2520

**Comment by MiguelLavigne** [comment]
Amazing thanks!

**Comment by ferdeli** [comment]
When will the 3.1.2 be released, thx?

**Comment by jongerrish** [comment]
We'll try and push a 3.1.2 tomorrow, but this won;t be in it.

**Comment by ferdeli** [comment]
@jongerrish gotcha ~ :(

**Comment by dominicthomas** [comment]
@jongerrish When is it likely there will be a fix? Thanks

**Comment by jongerrish** [comment]
I'd say by end of September. Fixing it properly requires quite a bit of
work but we'll get a hack in to get people unblocked.

Sorry about the delays we've got quite a bit of work to upgrade internally
first, and both me and Xian are out for a couple of weeks.

On Aug 17, 2016 9:09 AM, "Dominic Thomas" <notifications@github.com> wrote:

@jongerrish <https://github.com/jongerrish> When is it likely there will be
a fix? Thanks

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/robolectric/robolectric/issues/2315#issuecomment-240462189>,
or mute the thread

**Comment by Vandalko** [comment]
if someone else end here wondering why `abc_vector_test` image is not loading - here is the solution...

For API level >=24:
```java
  @Implements(className = "android.content.res.ResourcesImpl")
  public static class ShadowDrawableResourcesImpl extends ShadowResourcesImpl {
    public ShadowDrawableResourcesImpl() {
    }

    @TargetApi(21)
    @Implementation
    @Override
    public Drawable loadDrawable(Resources wrapper, TypedValue value, int id, Resources.Theme theme, boolean useCache) throws Resources.NotFoundException {
      try {
        return super.loadDrawable(wrapper, value, id, theme, useCache);
      } catch (Resources.NotFoundException e) {
        return new VectorDrawable();
      }
    }
  }
```

For API level 21, 22, 23:
To use it on API level <21 you would need to instantiate `VectorDrawableCompat` through reflections.
```java
  @Implements(Resources.class)
  public static class ShadowDrawableResources extends ShadowResources {
    public ShadowDrawableResources() {
    }

    @TargetApi(21)
    @Implementation
    @Override
    public Drawable loadDrawable(TypedValue value, int id) {
      try {
        return super.loadDrawable(value, id);
      } catch (Resources.NotFoundException e) {
        return new VectorDrawable();
      }
    }

    @TargetApi(21)
    @Implementation
    @Override
    public Drawable loadDrawable(TypedValue value, int id, Resources.Theme theme) throws Resources.NotFoundException {
      try {
        return super.loadDrawable(value, id, theme);
      } catch (Resources.NotFoundException e) {
        return new VectorDrawable();
      }
    }
  }
```

P.S. Why do I return `VectorDrawable` ?
Because of `android.support.v7.widget.AppCompatDrawableManager#checkVectorDrawableSetup`

It tries to load `abc_vector_test` and tests if it is an instance of correct class.

```java
    private void checkVectorDrawableSetup(@NonNull Context context) {
        if (mHasCheckedVectorDrawableSetup) {
            // We've already checked so return now...
            return;
        }
        // Here we will check that a known Vector drawable resource inside AppCompat can be
        // correctly decoded
        mHasCheckedVectorDrawableSetup = true;
        final Drawable d = getDrawable(context, R.drawable.abc_vector_test);
        if (d == null || !isVectorDrawable(d)) {
            mHasCheckedVectorDrawableSetup = false;
            throw new IllegalStateException("This app has been built with an incorrect "
                    + "configuration. Please configure your build for VectorDrawableCompat.");
        }
    }

    private static boolean isVectorDrawable(@NonNull Drawable d) {
        return d instanceof VectorDrawableCompat
                || PLATFORM_VD_CLAZZ.equals(d.getClass().getName());
    }
```

**Comment by dmitrikudrenko** [comment]
@Vandalko how to use your workaround?

**Comment by dmitrikudrenko** [comment]
@Vandalko your workaround doesn't work for me. Still receive ```android.content.res.Resources$NotFoundException``` error on ```AppCompatDrawableManager.checkVectorDrawableSetup```

**Comment by Vandalko** [comment]
@dmitrikudrenko I've recently been working with Robolectric 3.6.1 and that's what I use to workaround VectorDrawable problems:

```
package com.example.shadows;

import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.VectorDrawable;
import android.util.TypedValue;

import org.robolectric.annotation.Implementation;
import org.robolectric.annotation.Implements;
import org.robolectric.shadows.ShadowResourcesImpl;

import static android.os.Build.VERSION_CODES.N;

@Implements(className = "android.content.res.ResourcesImpl", inheritImplementationMethods = true, isInAndroidSdk = false, minSdk = N)
public class ShadowDrawableResourcesImpl extends ShadowResourcesImpl {
  public ShadowDrawableResourcesImpl() {
  }

  @Implementation
  @Override
  public Drawable loadDrawable(Resources wrapper, TypedValue value, int id, Resources.Theme theme, boolean useCache) throws Resources.NotFoundException {
    try {
      return super.loadDrawable(wrapper, value, id, theme, useCache);
    } catch (Exception e) {
      return new VectorDrawable();
    }
  }

  @Implementation
  @Override
  public Drawable loadDrawable(Resources wrapper, TypedValue value, int id, int density, Resources.Theme theme) {
    try {
      return super.loadDrawable(wrapper, value, id, density, theme);
    } catch (Exception e) {
      return new VectorDrawable();
    }
  }
}
```

And in your test class add:
```
@Config(shadows = {ShadowDrawableResourcesImpl.class}, sdk = 27)
```

**Comment by dmitrikudrenko** [comment]
@Vandalko still fails. My test - https://github.com/dmitrikudrenko/MDRXL/blob/tests/library/src/test/java/io/github/dmitrikudrenko/mdrxl/mvp/RxActivityTest.java
